### PR TITLE
feature:hrt358 - added novalet download

### PR DESCRIPTION
**Pull request by jasattilia**
## Link to JIRA ticket

Add a link to the JIRA ticket that the changes in this PR describe.

### *What is the problem we're trying to solve*

Add a summary of the problem and any additional context that helps explain what the issue we're trying to solve is.

### *What changes have we introduced*

Add descriptions of what changes were made to address the problem and provide a solution.

Include any links to commits that introduce significant additions of code if they help explain the process of coming to the solution. e.g Addition of test database setup, addition of first test and production class, removal of buggy code, etc.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [ ] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ ] Checked all code for possible refactoring
- [ ] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

File: HousingRegisterApi/V1/UseCase/GetNovaletExportUseCase.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            ICSVService csvService)
+        {
+            _logger = logger;
+            _gateway = gateway;
+            _csvService = csvService;
+        }
+
+        public async Task<FileExportResult> Execute(Guid id)
+        {
+            var application = _gateway.GetApplicationById(id);
```
Comment (tcmorris) [inline HousingRegisterApi/V1/UseCase/GetNovaletExportUseCase.cs:29]: Not sure why we would only want to export one application at a time? The export should contain all the applicants that need to bid on properties.

File: HousingRegisterApi/V1/UseCase/GetNovaletExportUseCase.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+            if (application == null)
+            {
+                return null;
+            }
+
+            var exportDataRow = new NovaletExportDataRow(application);
+            var bytes = await _csvService.Generate(exportDataRow, new CsvParameters()
+            {
+                IncludeHeaders = true
```
Comment (tcmorris) [inline HousingRegisterApi/V1/UseCase/GetNovaletExportUseCase.cs:40]: Default to `true` in the constructor as this is likely always desired.

File: HousingRegisterApi/V1/Services/CSVGeneratorService.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+            return await Task.Run(() =>
+            {
+                return GenerateBytes(array, csvParameters);
+            }).ConfigureAwait(false);
+        }
+
+        private static byte[] GenerateBytes<T>(IList<T> source, CsvParameters csvParameters)
+        {
+            MemoryStream ms = new MemoryStream();
```
Comment (tcmorris) [inline HousingRegisterApi/V1/Services/CSVGeneratorService.cs:35]: `MemoryStream` should be disposed as well?

File: HousingRegisterApi/V1/Services/CSVGeneratorService.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+using System.Threading.Tasks;
+using System.Linq;
+
+namespace HousingRegisterApi.V1.Services
+{
+    public class CSVGeneratorService : ICSVService
```
Comment (tcmorris) [inline HousingRegisterApi/V1/Services/CSVGeneratorService.cs:11]: From a naming point of view, I would go for `CsvGeneratorService` to separate Csv from Generator

File: HousingRegisterApi/V1/Services/CSVGeneratorService.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            headers = GetHeaders(typeof(T));
+            rowsOfValues = new List<List<object>>();
+
+            foreach (var entity in array)
+            {
+                rowsOfValues.Add(GetValues(entity));
+            }
+        }
+
+        private static string ConvertToString(object dataItem)
```
Comment (tcmorris) [inline HousingRegisterApi/V1/Services/CSVGeneratorService.cs:65]: In the past, I've used packages like [CsvHelper](https://joshclose.github.io/CsvHelper/) to handle this.

File: HousingRegisterApi/V1/UseCase/GetNovaletExportUseCase.cs (hunk 1 above)
Comment (jasattilia) [reply HousingRegisterApi/V1/UseCase/GetNovaletExportUseCase.cs]: yup, corrected.. i thought the export was per application.. now changed to an array of..

**Pull request merged**